File: recursion/standard-uses/bst_search.py
from BST_classes import Node, BST
                 
def bst_search(node, search_item):
    """Recursively searches a binary search tree. Returns a Boolean."""
    if search_item == node.data: 
        return True
    elif search_item > node.data and node.right != None:
        return bst_search(node.right, search_item)
    elif search_item < node.data and node.left != None:
        return bst_search(node.left, search_item)
    return False

def test_data(bst):
    """Inserts some test data into tree."""
    bst.insert('fred')
    bst.insert('elaine')
    bst.insert('george')
    bst.insert('harry')

def test(bst):
    """Carries out some basic tests."""
    start_node = bst.get_root() # Start search at root
    if bst_search(start_node, "aaron") != False:
        print("aaron - test failed")
    if bst_search(start_node, "harry") != True:
        print("harry - test failed")
    if bst_search(start_node, "zac") != False:
        print("zac - test failed")
    print("Tests complete")
    
       
if __name__ == '__main__':
    my_bst = BST()
    test_data(my_bst)  # Create some test data
    test(my_bst)  # Run test script

File: recursion/standard-uses/bst_search.cs
/*
Note: This file is designed to be copied out and compiled on your machine. 
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file. 

To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Compile the program
4. Run the program
*/

using System;

namespace isaac_code_samples
{
    class BSTSearchExample
    {
        // The Main method is the entry point for all C# programs
        static void Main(string[] args){
            BST myBST = new BST();
            TestData(myBST); // Create some test data
            Test(myBST); // Run test script
        }

        // A recursive search of a binary search tree
        static bool BSTSearch(Node node, string searchItem){
            int compare = String.Compare(searchItem, node.data, StringComparison.OrdinalIgnoreCase);
            if (searchItem == node.data){
                return true;
            } else if (compare > 0 && node.right != null){
                    return BSTSearch(node.right, searchItem);
            } else if (compare < 0 && node.left != null){
                return BSTSearch(node.left, searchItem);
            }
            return false;
        }

        // Inserts some data into tree
        static void TestData(BST bst){
            bst.Insert("fred");
            bst.Insert("elaine");
            bst.Insert("george");
            bst.Insert("harry");
        } 

        static void Test(BST bst){
            // Carries out some basic tests
            Node startNode = bst.GetRoot();
            if (BSTSearch(startNode, "aaron") != false){
                Console.WriteLine("aaron - test failed");
            }
            if (BSTSearch(startNode, "harry") != true){
                Console.WriteLine("harry - test failed");
            }
            if (BSTSearch(startNode, "zac") != false){
                Console.WriteLine("zac - test failed");
            }
            Console.WriteLine("Test complete");
        }        
    }
}